Incremental GC Forwarding Pointers

luc-blaeser opened the pull request:
### Incremental GC PR Stack
The Incremental GC is structured in three PRs to ease review:
1. https://github.com/dfinity/motoko/pull/3837
2. https://github.com/dfinity/motoko/pull/3831
3. https://github.com/dfinity/motoko/pull/3829 **<-- this PR**

# Incremental GC Forwarding Pointers

Support for forwarding pointers (Brooks pointer) to enable incremental moving (evacuating compacting) GC. 

Each object stores a forwarding pointer with the following properties:
* **Self-reference**: If the object resides at a valid location (i.e. not been relocated to another address), the forwarding pointer stores a reference to the object itself.
* **Single-level redirection**: If an object has been moved, the original object stores a pointer to the new object location. This implies that the data at the original object location is no longer valid.

Forwarding is only used during the evacuation phase and the updating phase of the incremental GC.

Indirection is at most one level, i.e. the relocation target cannot forward again to another location. The GC would need to update all incoming pointers before moving an object again in the next GC run.

Invariant: `object.forward().get_ptr() == object.get_ptr() || object.forward().forward().get_ptr() == object.forward().get_ptr()`.

## Changes

Changes made to the compiler and runtime system:
* **Header extension**: Additional object header space has been inserted for the forwarding pointer. Allocated objects forward to themselves.
* **Access indirection**: Every load, store, and object reference comparison effects an additional indirection through the forwarding pointer.

## Runtime Costs

Measuring the performance overhead of forwarding pointers in the GC benchmark using release mode (no sanity checks).

Total number of mutator instructions, average across all benchmark cases:

Configuration     | Avg. mutator instructions
------------------|---------------------------
No forwarding     | 1.61e10                   
With forwarding   | 1.80e10                   

**Runtime overhead of 12% on average.**

## Memory Costs

Allocated memory size, average across all GC benchmark cases, copying GCs:

Configuration     | Avg. final heap size
------------------|----------------------
No forwarding     | 306 MB
With forwarding   | 325 MB

**Memory overhead of 6% on average.**

## Testing

Extensive sanity checks for forwarding pointers were implemented and run in the separate PR (https://github.com/dfinity/motoko/pull/3546) containing the following sanity check code:
* **Indirection check**: Every derefencing of a forwarding pointer is checked whether the pointer is valid, and the invariant above holds.
* **Memory scan**: At GC time points of the existing copying or compacting GC, in regular intervals, the entire memory is scanned and all objects and pointers are verifying to be valid (valid forwarding pointer and plausible object tag).
* **Artificial forwarding**: For every created object, an artificial dummy object is returned that forwards to the real object. The dummy object stores zeroed content and has an invalid tag. This helps to verify that all heap accesses are correctly forwarded. Artificial forwarding disables the existing garbage collectors (due to the dummy objects not handled by the GCs) and performs memory scans at a defined frequency instead.

## Design Alternative 

* **Combining tag and forwarding pointer**: https://github.com/dfinity/motoko/pull/3904. This seems to be less efficient than the Brooks pointer technique with a runtime performance degrade of 27.5%, while only offering a small memory saving of around 2%.

## Reference

[1] R. A. Brooks. Trading Data Space for Reduced Time and Code Space in Real-Time Garbage Collection on Stock Hardware. ACM Symposium on LISP and Functional Programming, LFP'84, New York, NY, USA, 1984.

crusso commented on default.nix:
```diff
@@ -754,7 +754,7 @@ rec {
       rts
       base-src
-      base-tests
+      # base-tests # Re-enable when `RBTreeTestMore` has been scaled down
```
should be ok now, I hope.

luc-blaeser replied:
Indeed, I reenabled it. Thanks.

crusso commented on rts/motoko-rts-tests/src/gc.rs:
```diff
@@ -317,9 +325,13 @@ fn compute_reachable_objects(
     assert_eq!(read_word(heap, offset), TAG_ARRAY);
     offset += WORD_SIZE;
 
+    assert_eq!(read_word(heap, offset), make_pointer(table_addr as u32));
```
Is this checking the forwarding pointer points directly at self? Why can't it have been forwarded properly at this point?

luc-blaeser replied:
Good point. This is just temporarily in this PR which does not yet use the incremental GC. It will be refined in the higher PR.

crusso reviewed (commented):
Interim comments, will resume tomorrow.

luc-blaeser commented on rts/motoko-rts/src/memory/ic.rs:
```diff
@@ -19,6 +19,8 @@ pub(crate) static mut HP: u32 = 0;
 /// Heap pointer after last GC
 pub(crate) static mut LAST_HP: u32 = 0;
 
+pub(crate) static mut ALIGN: bool = false;
```
I would recommend so too. Only the copying GC is having this exception of non-alignment. I will change this.

luc-blaeser replied:
I believe the `ALIGN` variable is again removed in the higher PR, but the distinction is still made. I guess we could indeed unify this.

luc-blaeser commented on rts/motoko-rts/src/stream.rs:
```diff
@@ -14,13 +14,15 @@
-// - `tag` and `len` are blob metadata
+// - `obj header` contains tag and forwarding pointer
+// - `len` is in blob metadata
+// - 'padding' to align to 64-bit
```
Actually, it is critical, because the Rust compiler (according to C-memory representation) automatically aligns the next field `ptr64` to 64-bit. I only added this field for clarity. I took me a while to figure out that bug...

crusso replied:
> I took me a while to figure out that bug...
I bet!

luc-blaeser commented on rts/motoko-rts/src/types.rs:
```diff
@@ -387,9 +445,16 @@ pub const TAG_ARRAY_SLICE_MIN: Tag = 32;
 #[repr(C)] // See the note at the beginning of this module
 pub struct Obj {
     pub tag: Tag,
+    /// Forwarding pointer to support object moving in the incremental GC.
+    pub forward: Value,
```
Interesting idea. I will try this out. The only critical point is that certain artificial heap blocks, such as `OneWordFiller`, `FreeSpace`, and `FwdPtr` (used by the copying GC) would then have a different offset for the tag than the regular objects with the `Obj` header.

crusso replied:
Probably not worth it for now. I could well be that the compiler generates just as efficient code for our artificial cost model anyway. Maybe one to consider for the future, after merging the GC.

luc-blaeser replied:
**Update**: With the introduction of the mark bitmap, no linear partition memory scan is now performed by the incremental GC, and one could again consider swapping tag and forwarding pointer. Although, it could be fragile, if we would like to later use linear heap scan at some time.

luc-blaeser commented on rts/motoko-rts/src/types.rs:
```diff
@@ -256,63 +257,120 @@ impl Value {
+        debug_assert!(self.is_obj());
+        debug_assert!(self.get_ptr() as *const Obj != null());
+        let obj = self.get_ptr() as *const Obj;
+        (*obj).forward
```
I will try this out and measure the difference.

luc-blaeser replied:
**Update**: With the introduction of the mark bitmap and the separate RTS builds (https://github.com/dfinity/motoko/pull/3913), one could again consider swapping tag and forwarding pointer. I will put it on my backlog.

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -324,6 +325,7 @@ module E = struct
     named_imports = ref NameEnv.empty;
     built_in_funcs = ref NameEnv.empty;
     static_strings = ref StringEnv.empty;
+    object_pool = ref StringEnv.empty;
```
This is for pooling the shared static objects (every object except MutBox and dynamically created objects). Before they were handled as `static_strings` by simply comparing their memory state for lookup. Due to the forwarding pointer in its header, the state of these static objects now depends on their allocation location and this lookup is no longer working. This is why I separated this logic by a different lookup (taking only the tag and the payload).

crusso replied:
Might be worth a comment somewhere, if not described already.

luc-blaeser replied:
I agree. I added a comment at the declaration of `object_pool`.

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -1396,13 +1393,25 @@ module BitTagged = struct
+    compile_unboxed_const 1l ^^ (* true literal *)
+    G.i (Compare (Wasm.Values.I32 I32Op.Ne)) ^^
+    get_value ^^
+    compile_bitand_const 0x1l ^^
```
Thanks for the idea. I will try this.

crusso replied:
> Thanks for the idea. I will try this.

I'm just spotting optimization for the future. Don't feel obliged to do them before merging unless you think its a real win.

luc-blaeser replied:
Thanks, I will put it in my backlog.

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -1529,15 +1551,31 @@ module Tagged = struct
+  let alloc env size tag =
+    let (set_object, get_object) = new_local env "new_object" in
+    Heap.alloc env size ^^
+    set_object ^^ get_object ^^
+    compile_unboxed_const (int_of_tag tag) ^^
+    Heap.store_field tag_field ^^
+    get_object ^^ (* object pointer *)
+    get_object ^^ (* forwarding pointer *)
+    Heap.store_field forwarding_pointer_field ^^
```
This could be beneficial (since allocations are particularly performance-relevant). I will try this out.

crusso replied:
I'm just spotting optimizations for the future. Don't feel obliged to do them before merging unless you think its a real win.

luc-blaeser replied:
Thanks, I will put it in my backlog.

crusso commented on rts/motoko-rts/src/static_checks.rs:
```diff
@@ -17,18 +20,18 @@ const _: () = assert!(size_of::<usize>() == WORD_SIZE);
-const _: () = assert!(size_of::<Null>() == 1 * WORD_SIZE);
-const _: () = assert!(size_of::<Bits32>() == 2 * WORD_SIZE);
-const _: () = assert!(size_of::<Bits64>() == 3 * WORD_SIZE);
+const _: () = assert!(size_of::<Obj>() == HEADER_SIZE);
```
No, its fine. I just got confused.

crusso commented on src/codegen/compile.ml:
```diff
@@ -1465,19 +1474,32 @@ module BitTagged = struct
 
      The tag is a word at the beginning of the object.
 
+     They forwarding pointer supports object moving in the incremental garbage collection.
```
```suggestion
     The (skewed) forwarding pointer supports object moving in the incremental garbage collection. 
```

crusso commented on src/codegen/compile.ml:
```diff
@@ -1465,19 +1474,32 @@ module BitTagged = struct
+     │ tag  │ fwd ptr │ ...
+     └──────┴─────────┴──
+
      All tagged heap objects have a size of at least two words
```
```suggestion
     All tagged heap objects have a size of at least three words
```
?

luc-blaeser replied:
Thank you, I indeed should refine this comment. There is a special exception with `NULL` that now has a size of two words and even occupied only one word before this PR. However, as it resides in the static heap, there is no problem with the indirection during the copying GC.

crusso commented on src/codegen/compile.ml:
```diff
@@ -8927,20 +9075,21 @@ and compile_prim_invocation (env : E.t) ae p es at =
        fields, effectively arriving at the desired element *)
     G.i (Binary (Wasm.Values.I32 I32Op.Add)) ^^
-    Arr.load_field 0l                (* loads the element at the byte offset *)
+    Heap.load_field Arr.header_size (* loads the element at the byte offset *)
```
Is this to avoid double forwarding in Arr.load_field?

luc-blaeser replied:
Yes, but you are right: It is not consistent with the other places where there is still duplicated forwarding.

crusso commented on src/codegen/compile.ml:
```diff
@@ -5557,16 +5705,16 @@ module MakeSerialization (Strm : Stream) = struct
         write_byte env get_data_buf (compile_unboxed_const 1l) ^^
-        get_x ^^ Arr.load_field 0l ^^ write env (Obj (Actor, [])) ^^
-        get_x ^^ Arr.load_field 1l ^^ write env (Prim Text)
+        get_x ^^ Arr.load_field env 0l ^^ write env (Obj (Actor, [])) ^^
+        get_x ^^ Arr.load_field env 1l ^^ write env (Prim Text)
```
Duplicated forwarding (if we care).

luc-blaeser replied:
We could use this logic. But I am not sure whether there is a performance gain (or even degrade as for `Arr.idx`). 
```
        get_x ^^ Tagged.load_forwarding_pointer env ^^ set_x ^^
        inc_data_size (compile_unboxed_const 1l) ^^ (* one byte tag *)
        get_x ^^ Heap.load_field Int32.(add 0l Arr.header_size) ^^ size env (Obj (Actor, [])) ^^
        get_x ^^ Heap.load_field Int32.(add 1l Arr.header_size) ^^ size env (Prim Text)
```

crusso replied:
Probably not worth it then. This code is rarely used (higher-order Candid)

crusso reviewed (commented):
Took a pass through compile.ml, looks  good. Also spotted some unrelated optimization opportunities (#3852, #3851)

I guess the hard part is spotting any forwarding that has been missed.... might look for some of those tomorrow.

luc-blaeser commented:
> #3851

This is great. Thank you very much. This is a lot of work. Excellent ideas for optimizations (thanks for registering issue tickets for them).

I also believe that the risky thing is having missed forwarding resolution. Maybe it helps that I also study this alternative idea of using a central forwarding table that would probably show a missing indirection more reliably (as the references are always synthetic small values that have to be resolved).

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -553,6 +559,13 @@ module E = struct
 
+  let object_pool_find (env: t) (key: string) : int32 option =
+    StringEnv.find_opt key !(env.object_pool)
+    
```
Thank you. Improved this together with a documentation of `object_pool`.

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -553,6 +559,13 @@ module E = struct
+  let object_pool_add (env: t) (key: string) (ptr : int32)  : unit =
+    env.object_pool := StringEnv.add key ptr !(env.object_pool);
+    ()
+    
```
Thank you. Improved this together with a documentation of `object_pool`.

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -1529,15 +1555,31 @@ module Tagged = struct
-  let load =
+    
+  let load_tag env =
+    load_forwarding_pointer env ^^
```
This is an interesting question. I was also first thinking of only resolving it in sanity check mode. However, I believe it could be relevant during stabilization where the tags are changed to store the serialization sharing status (`StableSeen`, serialization offset). As serialization may start on non-completed evacuation and see old references, it would be important to also forward the tag accesses (load and store). Maybe, this could be simplified (or only done during serialization)?

crusso replied:
Oh, I forgot about that hack during stabilization, well observed!

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -1529,15 +1555,31 @@ module Tagged = struct
+  let load_forwarding_pointer env =
+    Heap.load_field forwarding_pointer_field
+
+  let store_tag env tag =
+    load_forwarding_pointer env ^^
```
Also here, it could be different during stabilization that changes the tags (only in the forwarded location)...

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -2155,36 +2274,35 @@ module Float = struct
      treated opaquely by the RTS. We'll introduce a separate tag when the need of
      debug inspection (or GC representation change) arises.
 
+     The object header includes the object tag (Bits64) and the forwarding pointer.
```
Indeed, it would be valuable to have a precise dynamic type metadata for the heap (and also for the stack).

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -3583,14 +3724,15 @@ module Arr = struct
       (* Check the upper bound *)
       get_idx ^^
-      get_array ^^ Heap.load_field len_field ^^
+      get_array ^^ len env ^^
       G.i (Compare (Wasm.Values.I32 I32Op.LtU)) ^^
       E.else_trap_with env "Array index out of bounds" ^^
 
       get_idx ^^
       compile_add_const header_size ^^
       compile_mul_const element_size ^^
       get_array ^^
+      Tagged.load_forwarding_pointer env ^^
```
Indeed, this could offer nice optimization potential considering that array element accesses are frequent. I will do this and measure the performance gain.

luc-blaeser replied:
I tried the change below to avoid multiple forwarding resolution. Interestingly, it is slower with around 0.5% performance degrade for array-intense GC benchmark cases...

```
      let set_array = G.setter_for get_array in
      get_array ^^ Tagged.load_forwarding_pointer env ^^ set_array ^^
      (* No need to check the lower bound, we interpret idx as unsigned *)
      (* Check the upper bound *)
      get_idx ^^
      get_array ^^ Heap.load_field len_field ^^
      G.i (Compare (Wasm.Values.I32 I32Op.LtU)) ^^
      E.else_trap_with env "Array index out of bounds" ^^

      get_idx ^^
      compile_add_const header_size ^^
      compile_mul_const element_size ^^
      get_array ^^
      G.i (Binary (Wasm.Values.I32 I32Op.Add))
      ```

crusso replied:
Joachim just opened a PR to remove the bounds check anyway #3853, which would remove the redundant forwarding anyway.

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -7684,9 +7828,9 @@ module FuncDec = struct
       (* The callee *)
-      get_meth_pair ^^ Arr.load_field 0l ^^ Blob.as_ptr_len env ^^
+      get_meth_pair ^^ Arr.load_field env 0l ^^ Blob.as_ptr_len env ^^
       (* The method name *)
-      get_meth_pair ^^ Arr.load_field 1l ^^ Blob.as_ptr_len env ^^
+      get_meth_pair ^^ Arr.load_field env 1l ^^ Blob.as_ptr_len env ^^
```
I will improve this. Thanks. Not sure how frequent this is executed.

crusso replied:
I wouldn't bother, tbh. I was just blindly pointing them out.

luc-blaeser commented on src/codegen/compile.ml:
```diff
@@ -7755,9 +7899,9 @@ module FuncDec = struct
       (* The callee *)
-      get_meth_pair ^^ Arr.load_field 0l ^^ Blob.as_ptr_len env ^^
+      get_meth_pair ^^ Arr.load_field env 0l ^^ Blob.as_ptr_len env ^^
       (* The method name *)
-      get_meth_pair ^^ Arr.load_field 1l ^^ Blob.as_ptr_len env ^^
+      get_meth_pair ^^ Arr.load_field env 1l ^^ Blob.as_ptr_len env ^^
```
I tried the code below but not noticed a performance benefit (but maybe my benchmark cases have to few message calls).

```
      get_meth_pair ^^ Tagged.load_forwarding_pointer env ^^ set_resolved ^^
      (* The callee *)
      get_resolved ^^ Heap.load_field Int32.(add 0l Arr.header_size) ^^ Blob.as_ptr_len env ^^
      (* The method name *)
      get_resolved ^^ Heap.load_field Int32.(add 1l Arr.header_size) ^^ Blob.as_ptr_len env ^^
```

crusso reviewed (commented):
Would it make sense to, in Heap.load_field n, when n<>forwarding pointer, to check that the forwarding pointer points to self? Or introduce a Tagged.load_field that does the check (and use it where appropriate).

Only in sanity checking runs, of course.

On the other hand, I eye-balled all Heap.load_fields and they appear to only act on resolved pointers.

luc-blaeser commented on rts/motoko-rts/src/types.rs:
```diff
@@ -387,9 +445,16 @@ pub const TAG_ARRAY_SLICE_MIN: Tag = 32;
 impl Obj {
+    /// Check whether the object's forwarding pointer refers to a different location.
+    pub unsafe fn is_forwarded(self: *const Self) -> bool {
+        (*self).forward.get_ptr() != self as usize
```
The difficulty is that `OneWordFiller`, `FreeSpace` and `FwdPtr` only have a tag and no forwarding pointer. So `Obj` is an extension of that layout. Currently, these special blocks are only used in the copying GC and also in `Stream` (creating free space). If we would remove those, swapping the tag and forwarding pointer would be easy.

luc-blaeser replied:
**Update**: With the introduction of the mark bitmap and the separate RTS builds (https://github.com/dfinity/motoko/pull/3913), one could again consider swapping tag and forwarding pointer. I will put it on my backlog.

ulan commented on rts/motoko-rts/src/types.rs:
```diff
@@ -256,63 +257,120 @@ impl Value {
+    pub unsafe fn check_forwarding_pointer(self) {
+        debug_assert!(
+            self.forward().get_ptr() == self.get_ptr()
+                || self.forward().forward().get_ptr() == self.forward().get_ptr()
```
In the current PR, an even stronger invariant `self.forward().get_ptr() == self.get_ptr()` should hold, right?

luc-blaeser replied:
Indeed. It is a preparation for the third PR (probably a left-over when I split the PRs). I will update this for this PR.

ulan commented on rts/motoko-rts/src/gc/copying.rs:
```diff
@@ -23,7 +23,7 @@ unsafe fn copying_gc<M: Memory>(mem: &mut M) {
     copying_gc_internal(
         mem,
-        ic::get_heap_base(),
+        ic::get_aligned_heap_base(),
```
What is the motivation for this change?

luc-blaeser replied:
It is actually unrelated to this PR but only a unification of the heap alignment, mentioned in: 
https://github.com/dfinity/motoko/pull/3829#discussion_r1119110197
Not sure whether I should revert this to keep the PR more coherent?

ulan commented on rts/motoko-rts/src/types.rs:
```diff
@@ -367,7 +425,7 @@ pub const TAG_CLOSURE: Tag = 11;
 pub const TAG_VARIANT: Tag = 15;
 pub const TAG_BLOB: Tag = 17;
-pub const TAG_FWD_PTR: Tag = 19;
+pub const TAG_FWD_PTR: Tag = 19; // Only used by the copying GC - not to be confused with forwarding pointer in the header used for incremental GC.
```
> "Only used by the copying GC"

A more precise statement would be "Only used by non-incremental GCs" ? As a future cleanup, would it make sense to use the new forwarding fields in the old GCs as well?

luc-blaeser replied:
Currently, this special tag/record is only used by the copying GC (not by the other GCs). Maybe we should in future even revise which GCs should stay in the runtime system. We could indeed use the forwarding pointer in the copying GC - not sure whether I should already unify it now. I will keep it in my backlog.

ulan reviewed (approved):
Looks great! Thanks for extracting this change into a separate PR.
My comments are mostly nits and questions. Please feel free to postpone addressing the nits in subsequent PR to simplify merging.

luc-blaeser commented on rts/motoko-rts/src/types.rs:
```diff
@@ -256,63 +257,120 @@ impl Value {
+    /// Check whether the object's forwarding pointer refers to a different location.
+    pub unsafe fn is_forwarded(self) -> bool {
+        self.check_forwarding_pointer();
+        self.forward().get_ptr() != self.get_ptr()
```
I also thought about this. This is very good question. My motivation was to simplify the changes in the compiler, as the forwarding pointer has the same encoding as the non-forwarded pointers. There are several places in the compiler where it passes (skewed) pointers (especially also in the quite complex serialization logic). So, I indeed traded simplicity for performance. But this could be something for the backlog to optimize.

luc-blaeser commented:
> Looks great! Thanks for extracting this change into a separate PR. My comments are mostly nits and questions. Please feel free to postpone addressing the nits in subsequent PR to simplify merging.

Thank you very much, Ulan, for your valuable feedback and review points.

crusso commented on rts/motoko-rts/src/types.rs:
```diff
@@ -256,63 +257,120 @@ impl Value {
     pub unsafe fn tag(self) -> Tag {
-        debug_assert!(self.get().is_ptr());
-        (self.get_ptr() as *mut Obj).tag()
+        *(self.get_ptr() as *const Tag)
```
I notice that many old calls to `.tag()` have been replaced by `.tag`. Is that just more efficient or some other reason?

luc-blaeser replied:
Thanks for asking this. The changes from `.tag()` to `.tag` are actually a remainder from earlier changes, where I added a forwarding check inside the `tag()` function. This was not compatible with the threading technique used in the compacting and generational GC. This is why I changed the code there to the low-level accesses. I will revert those superfluous changes. 

Actually, with the introduction of the mark bitmap, we could again consider swapping forwarding pointer and tag but it could be fragile if we still would like to perform a linear heap scan at some time... Maybe I can try in a separate PR to see how big the performance benefit would be.

crusso commented on rts/motoko-rts/src/types.rs:
```diff
@@ -262,57 +263,111 @@ impl Value {
+    /// Resolve forwarding if the value is a pointer. Otherwise, return the same value.
+    pub unsafe fn forward_if_possible(self) -> Value {
+        if self.is_ptr() && self.get_ptr() as *const Obj != null() {
+            // Ignore null pointers used in text_iter.
```
Why does text_iter use `null`? I wonder if this edge case could lead to bugs elsewhere.

luc-blaeser replied:
I assume it is an exception and could be improved by using a reference to the `Null` object. Maybe we should do this is in a separate PR? (Currently, I do not see a bug implication elsewhere but I had to consider this special case during GC work.)

crusso replied:
Sure no reason to do it here.

crusso reviewed (approved):
Ok, I think I need to stop starting a this code ;->

I mentioned it offline yesterday, but I we keep maintaining separate GCs and don't want to pay the overhead of the forwarding pointer for GCs that don't use them, then maybe we could consider storing the forwarding pointer at offset -1 of the heap objects that have them, removing it from the Obj struct and ensuring the remaining object layout, at positive offsets, is the same for all GCs. Hopefully the code for resolving forwarding pointers can just omit a nop for GC that don't require forwarding but much of the rest of the code will remain unchanged.

But maybe Gabor's suggesting of merging tag and forwarding pointer is even better, if it can be made to work.

luc-blaeser commented on rts/motoko-rts/src/gc/generational/sanity_checks.rs:
```diff
@@ -52,25 +52,27 @@ unsafe fn verify_heap(limits: &Limits) {
-            },
-        );
-        pointer += object_size(current as usize).to_bytes().as_usize();
+        if Value::from_ptr(pointer).is_obj() {
```
There is a check in the `incremental/sanity_check.rs` in the higher PR (because I thought to only enable forwarding with that GC mode) that checks that no forwarding is used when the GC is not active (completed updating). There are additional checks in `types.rs` with `check_forwarding_pointer()` that the forwarding is maximum one indirection.

luc-blaeser commented:
> Ok, I think I need to stop starting ta this code ;->
> 
> I mentioned it offline yesterday, but if we keep maintaining separate GCs and don't want to pay the overhead of the forwarding pointer for GCs that don't use them, then maybe we could consider storing the forwarding pointer at offset -1 of the heap objects that have them, removing it from the Obj struct and ensuring the remaining object layout, at positive offsets, is the same for all GCs (and similar to what it was before adding forwarding pointers). Hopefully the code for resolving forwarding pointers can just omit a nop for GCs that don't require forwarding but much of the rest of the code will remain unchanged. Object references would always point at the tag word (which was something that didn't work with my suggestion of swapping tag and forwarding pointer fields in Obj, I think).
> 
> But maybe Gabor's suggesting of merging tag and forwarding pointer is even better, if it can be made to work cheaply. I worry that might not work well with the threading of the compacting/generational gc, but maybe it's fine.

I agree. I think this could be solved by introducing a separate RTS build for the incremental GC which uses the forwarding pointer header layout (independent of the offset direction, -1 or +1). The compiler could probably also quite easily be adjusted to skip the forwarding pointer on compilation for non-incremental GC. 

I was thinking about the negative offset which is very interesting. The problem I see is the same as for swapping tag and  forwarding pointer. As we l have the special `FreeBlock` and `OneWordFiller` blocks without forwarding pointer, a heap scan during evacuation could not easily decide the block type when looking at the first word of the block: It could be the forwarding pointer (negative offset from perspective of regular objects) or a tag (offset 0) for special blocks. It could probably still be solved (with the same idea as Gabor mentioned that tag and forwarding pointer could be distinguished - with some exceptions).

I think I will first explore the idea of Gabor of merging the forwarding pointer and tag and measure the performance difference. If that approach is working well, we could simply eliminate the extra header word for the incremental GC. Otherwise, splitting the RTS in different GC versions and introducing conditional header layouts in the compiler could be viable.

luc-blaeser commented:
> I think, if you agree, that I should continue on separating the RTS builds and having a compiler switch to not impose the header extension costs during classical GC modes?

**Latest Update:**
I implemented the separate RTS builds with the conditional forwarding pointer header extension only for the incremental GC: https://github.com/dfinity/motoko/pull/3913

With this change, there should be no relevant performance degrade (memory or instructions) for the classical GCs while integrating incremental GC support at the same time.

crusso reviewed (approved):
Thanks for cleaning up the max-stack tests!

luc-blaeser commented:
> Thanks for cleaning up the max-stack tests!

Thank you also very much for reviewing!
